pkg/csi_driver/multishare_ops_manager.go

```diff
@@ -17,27 +17,550 @@ limitations under the License.
+	}
+
+	// verify instance is ready.
+	instanceReady, err := m.verifyNoRunningInstanceOps(ctx, instanceSCPrefix, w.share.Parent)
+	if err != nil {
+		return nil, status.Errorf(codes.Aborted, "Instance %q check error: %v", w.share.Parent.Name, err)
+	}
+	if !instanceReady {
+		return nil, status.Errorf(codes.Aborted, "Instance %q not ready", w.share.Parent.Name)
+	}
+
+	// Verify share is ready.
+	shareReady, err := m.verifyNoRunningShareOp(ctx, instanceSCPrefix, w.share)
+	if err != nil {
+		return nil, status.Errorf(codes.Aborted, "Share %q check error: %v", w.share.Name, err)
+	}
+	if !shareReady {
+		return nil, status.Errorf(codes.Aborted, "Share %q not ready", w.share.Name)
+	}
+
+	switch w.opType {
+	case util.ShareCreate:
+		// validate no entry in cache for the share
+		item := m.cache.GetShareCreateOp(instanceSCPrefix, w.share.Name)
+		if item != nil {
+			return nil, status.Errorf(codes.Aborted, "Share %q not ready, found op %q", w.share.Name, item.OpName)
+		}
+
+		op, err := m.cloud.File.StartCreateShareOp(ctx, w.share)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
+		}
+
+		// Start op and update cache
+		instanceHandle, err := file.GetMultishareInstanceHandle(w.share.Parent)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
+		}
+		m.cache.AddShareCreateOp(instanceSCPrefix, w.share.Name, util.ShareCreateOpInfo{
+			InstanceHandle: instanceHandle,
+			OpName:         op.Name,
+		})
+		w.opName = op.Name
+	case util.ShareExpand:
+		// validate no entry in cache for the share
+		item := m.cache.GetShareOp(instanceSCPrefix, util.CreateShareKey(w.share.Parent.Project, w.share.Parent.Location, w.share.Parent.Name, w.share.Name))
+		if item != nil {
+			return nil, status.Errorf(codes.Aborted, "Share %q not ready, found op %q", w.share.Name, item.Name)
+		}
+
+		// Start op and update cache
+		op, err := m.cloud.File.StartResizeShareOp(ctx, w.share)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
+		}
+		m.cache.AddShareOp(instanceSCPrefix, util.CreateShareKey(w.share.Parent.Project, w.share.Parent.Location, w.share.Parent.Name, w.share.Name), util.OpInfo{Name: op.Name, Type: util.ShareExpand})
+		w.opName = op.Name
+	case util.ShareDelete:
+		// validate no entry in cache for the share
+		item := m.cache.GetShareOp(instanceSCPrefix, util.CreateShareKey(w.share.Parent.Project, w.share.Parent.Location, w.share.Parent.Name, w.share.Name))
+		if item != nil {
+			return nil, status.Errorf(codes.Aborted, "Share %q not ready, found op %q", w.share.Name, item.Name)
+		}
+
+		// Start op and update cache
+		op, err := m.cloud.File.StartDeleteShareOp(ctx, w.share)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
+		}
+		m.cache.AddShareOp(instanceSCPrefix, util.CreateShareKey(w.share.Parent.Project, w.share.Parent.Location, w.share.Parent.Name, w.share.Name), util.OpInfo{Name: op.Name, Type: util.ShareExpand})
```

Comment:
the type should be `shareDelete` here.
Merge this case with `ShareExpand` and pass in w.OpType in instead

Reply:
good catch. Changed to use optype. keeping separate case for readbility and since the expand and delete use different functions StartResizeShareOp() / StartDeleteShareOp().